level_2/tests/semantic_tests.py

```diff
@@ -0,0 +1,119 @@
+import os
+import openai
+from deepeval.metrics.factual_consistency import assert_factual_consistency
+import dotenv
+dotenv.load_dotenv()
+from level_2.level_2_pdf_vectorstore__dlt_contracts import Memory
+openai.api_key = os.getenv("OPENAI_API_KEY", "")
+from deepeval.metrics.overall_score import assert_overall_score
+import json
+from deepeval.metrics.overall_score import OverallScoreMetric
+
+# Write a sample ChatGPT function
+
+
+async def main():
+    async def generate_context(query: str='bla', context:str=None):
+        memory = Memory(user_id="TestUser")
+
+        await memory.async_init()
+
+
+        memory_loaded = await memory._fetch_semantic_memory(observation=query, params=None)
+
+        if memory_loaded:
+            return memory_loaded["data"]["Get"]["SEMANTICMEMORY"][0]["text"]
+        else:
+            params = {
+                "version": "1.0",
+                "agreement_id": "AG123456",
+                "privacy_policy": "https://example.com/privacy",
+                "terms_of_service": "https://example.com/terms",
+                "format": "json",
+                "schema_version": "1.1",
+                "checksum": "a1b2c3d4e5f6",
+                "owner": "John Doe",
+                "license": "MIT",
+                "validity_start": "2023-08-01",
+                "validity_end": "2024-07-31",
+            }
+            loader_settings =  {
+            "format": "PDF",
+            "source": "url",
+            "path": "https://www.ibiblio.org/ebooks/London/Call%20of%20Wild.pdf"
+            }
+            load_jack_london = await memory._add_semantic_memory(observation = query, loader_settings=loader_settings, params=params)
+            memory_loaded = await memory._fetch_semantic_memory(observation=query, params=None)
+            return memory_loaded["data"]["Get"]["SEMANTICMEMORY"][0]["text"]
```

Comment:
just cleaner code - do the if inversely - create the vector if it doesn't exists and then do the fetch and return once instead of twice

Reply:
not sure i get this one. any pseudo code to understand it better?